RFD 78: Login Rules

nklaassen opened the pull request:
Issue: https://github.com/gravitational/teleport/issues/11042

This RFD outlines the design for the new login rules feature.

nklaassen committed:
first draft of Role Variables RFD

Tener commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+      select: "{{external.email}}"
+      expansions:
+        # nic should be able to login as root
+        - match: '^nic@goteleport\.com$'
```
I wonder if we are leaning too heavily on regular expressions here. While powerful, they have limitations in expressiveness and are hard to read; this makes role security auditing harder than it should be. Perhaps we can use expression language we support for traits, with additional extensions as needed?

nklaassen replied:
What do you think this would look like with an expression language?

Tener replied:
There are various ways to cut it, but we should aim for a clean expression of common patterns. So, what common use cases would we like to see?

nklaassen replied:
I have to say, I think regular expressions are a pretty powerful solution here. Most people are pretty familiar with them.

I can't even really think of what this would like like with an expression language, but I think you would mostly just want to do regex-like matching in that language. If you have a counter example I'd be interested to see it.

One "killer feature" of regex's here, that perhaps I haven't highlighted enough, is the ability to use capture groups to use parts of the input in the output value, like below. I think this is basically a necessary feature here and I'm not sure of a better way to enable it than regular expressions.
```
- match: '^env-(\w+)$'
  out: '$1'
```

Tener replied:
An expression language is a big thing to design and implement, so perhaps regexes offer a safe path forwards. Nevertheless, I'd consider using something like CEL (https://github.com/google/cel-spec, https://github.com/google/cel-go).

nklaassen replied:
I've added examples under the alternatives section written with the `predicate` language as well as CEL.

It's an interesting solution and it's pretty fun to experiment with, but it does feel complex and I can imagine fighting with it as a user. Not sure if what I came up with is similar to what you were thinking.

zmb3 commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+Teleport admins don't always have control over their org's OIDC/SAML provider.
+
+Role variables will simplify role evaluation, keep logic in one place, and
+minimize repetition.
```
Can we add a sample use case that illustrates something that is not possible without role variables?

nklaassen replied:
A lot is possible with enough roles and creative/repetitive use of `regex.replace`, variables will be great for minimizing repetition though. I will add some more examples.

nklaassen replied:
Added a good example in [d41536b](https://github.com/gravitational/teleport/pull/14977/commits/d41536b4df8e83de90de64ae75aebb5659655a14)

zmb3 commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+      # "select" is a string value which will be used as the input to all
+      # possible expansion matchers. It can (and usually will) contain a trait
+      # which will be expanded before passing as input.
+      select: "{{external.email}}"
```
I guess the ship has sailed with the `{{` `}}` syntax since we already use it for traits. I wish we would have gone with something different, because if you forget to quote it you get invalid YAML.

nklaassen replied:
Yeah I was planning to just use the regular trait expansion here, but we could make it so you give the name only, like `select: external.email`. That might make it more clear that there can only be 1, and and prefix/suffix is pointless anyway, I will probably switch to that

nklaassen replied:
Never mind, I will probably leave the `{{}}` syntax as-is to keep things simple and consistent with other uses of traits.

zmb3 commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+
+### What happens when the same variable name is defined in two different roles?
+
+This is fine, until a user tries to login with both roles, in which case the
```
That was going to be my suggestion as well.

nklaassen replied:
Yeah I like this idea.

nklaassen replied:
hmm, big downside here is that we use to Go parser which does not understand hyphens in the middle of identifiers, it thinks you are trying to do math with a minus sign. So roles like `node-users` wouldn't work, you'd have to rename to `node_users`...

nklaassen replied:
made this change

Tener replied:
Not sure if the parser would accept it, but maybe we can use string literals as a workaround? E.g. `vars."node-users".allow_env`, or something in that vein. Not great, but maybe it will allow to reference roles with dashes?

nklaassen replied:
I realized that the index syntax also works for this out of the box, i.e. you can do `{{var["role-name"]["variable-name"]}}` as necessary. Expanded on this in the doc.

nklaassen commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+
+To prevent confusion about variables intermittently being defined, we can
+enforce no variable definition can depend on variables defined in a different
+role. But I think it's simple and useful enough to allow variables to depend on
```
I was avoiding a topological sort just for simplicity's sake. If someone adds a role which creates a cycle, that could also break logins. Would need to load all roles, then parse (without traits) and topologically sort all variables during role creation to make sure a cycle is never created. It's probably doable though, I'll have another look at it.

Tener replied:
I used https://github.com/stevenle/topsort in the past, you may find it useful too.

nklaassen replied:
This is no longer relevant not that variables will be local to the role which defines them

nklaassen commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+      # "select" is a string value which will be used as the input to all
+      # possible expansion matchers. It can (and usually will) contain a trait
+      # which will be expanded before passing as input.
+      select: "{{external.email}}"
```
Our parser doesn't support multiple traits in a single string, I can't really think of a usecase where one would want that, if there is one you could probably use an intermediate variable.

nklaassen committed:
scope variables with role name

nklaassen committed:
fix typo

nklaassen commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+
+When fetching roles, we will detect if a variable is being redefined, and
+return an error.
+This will block login with two roles which define the same variable.
```
updated to scope variables by role

nklaassen committed:
expand on variable referencing options

nklaassen commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,176 @@
+This is especially true given Go's lack of support for negative-lookahead
+regular expressions.
+
+I'm not quite sure which is the best route to take here and am quite open to
```
I suppose I could also just add a `not_match: <regex>` field to allow negative matches.

nklaassen replied:
I have updated the syntax so that each possible value can have its own unique input, with a `match` and `not_match` expression, and this question is no longer relevant.

nklaassen committed:
switch to list of values with unique inputs

this also removes the question of whether to take the first or all
matches: each possible value is considered on its own, and may contain a
not_match expression

nklaassen committed:
fix typo

nklaassen committed:
add compelling example

klizhentas commented on rfd/0078-role-variables.md:
````diff
@@ -0,0 +1,264 @@
+      env: ["{{vars.node_users["allow-env"]}}"]
+```
+
+These variables could be used in another role without redefinition:
````
Let's keep variables role-local to avoid confusing behavior. Today, each role is self-sufficient and it's update is atomic. Role-local variables will keep the same guarantee. The downside is that some roles will have to declare the same variables many times. But I think it's OK for now.

klizhentas commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,264 @@
+or else the parser will not be able to handle it.
+The same is true for traits today.
+
+Variables can be referenced in the role which defines them, or in any other
```
Let's keep variables role-local. This adds some inconveniences, but has major upside in simplicity of evaluation and the lack of ambiguity. Linter can validate any role just by looking at it, and having no other context.

klizhentas commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,264 @@
+In order for a variable to be defined, the user must always posess the role
+which defines it.
+
+If the user does not have the role which defines the variable, or a typo is made,
```
When variables are node-local, we can avoid this problem by returning error when creating a role referencing unknown variable.

klizhentas reviewed (changes requested):
Thanks for writing down this RFD. 

Overall looks good, but let's keep variables role-local. We would avoid a lot of issues with global variables:

* Variable scope is unclear. 
* One role can define variable that is visible in other variables.
* A variable can overwrite logic in another role. 
* There is no collision protection mechanism or mechanism to detect references to undefined variables.

klizhentas reviewed (commented):
Alternatively, we can go back to older idea of using traits transformations. Consider this:


```yaml
kind: trait
version: v5
metadata:
  name: groups
spec:
  values:
        # An input will expand to a list of N inputs if the trait is a list of size N.
        # Any input value which matches the `match` regex will contribute one
        # output value, which may contain regex captures from that specific input.
        #
        # Here [devs, env-staging, env-prod] maps to [staging, prod].
        - input: "{{external.groups}}"
          match: '^env-(\w+)$'
          out: ['$1']
        # If any of the N input values match the `not_match` regex, none of the
        # values will be used.
        #
        # Here [devs] maps to [dev], but [devs, contractors] maps to []
        - input: "{{external.groups}}"
          match: '^devs$'
          not_match: 'contractors'
          out: ["dev"]
```

This follows the same logic, but instead applies the trait transformation at the time of login and embeds it in the cert. This trait can be then referenced in any role as usual. 

Let's discuss either role-local variables or traits transformations next week. @zmb3 @nklaassen

Tener commented:
> Let's discuss either role-local variables or traits transformations next week.

My 2c: we could use CEL (https://github.com/google/cel-spec, https://github.com/google/cel-go) for matching and transforming parts. This gives us a stable base to build upon without reinventing the wheel but still allowing for extensibility. I know we have our eyes on https://github.com/gravitational/predicate, but CEL appears to be more feature-rich: macros and object construction look particularly interesting.

nklaassen commented:
@klizhentas Thanks for your review.

If we will put the variables definitions in the role spec, I agree it makes sense to make them role-local only.

I like the trait transformations idea, I had thought of embedding the variables in the cert as a trait but didn't consider they could be defined separately from the role, I think your suggestion could offer pretty good semantics and would make the variables reusable across roles and potentially trusted clusters (something to consider at least.)

I will consider these some more, and I have sent a calendar invite to discuss this with you and @zmb3

nklaassen committed:
fix typos

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

nklaassen committed:
make variables role-local only

nklaassen committed:
make variables role-local only pt. 2

nklaassen committed:
describe trait transformation alternative

nklaassen committed:
describe CEL alternative

klizhentas commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,358 @@
+  name: node_users
+spec:
+  vars:
+    - name: "logins"
```
Let's not introduce a language that is almost like predicate, but a bit different. Instead, let's focus on investing in predicate.

klizhentas commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,358 @@
+    # output value, which may contain regex captures from that specific input.
+    #
+    # Here [devs, env-staging, env-prod] maps to [staging, prod].
+    - input: "{{external.groups}}"
```
if input is always a variable, may be we don't need any brackets or other syntax?

klizhentas commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,358 @@
+    # values will be used.
+    #
+    # Here [devs] maps to [dev], but [devs, contractors] maps to []
+    - input: "{{external.groups}}"
```
We need to think about the syntax, but overall I think `trait` solves our problem better than embedded role variables

nklaassen committed:
add predicate language example

klizhentas commented:
@Tener CEL looks interesting. However, we are already using predicate in Teleport elsewhere - all roles, OIDC connectors, search expressions. For that reason sticking with predicate is more reasonable choice for now. We could re-evaluate choosing a different language, but we should consider refactoring the entire Teleport code-base then.

klizhentas commented on rfd/0078-role-variables.md:
````diff
@@ -0,0 +1,439 @@
+
+The syntax could be similar to the proposed role variable syntax:
+
+```yaml
````
@nklaassen @zmb3 some thoughts ahead of our conversation today:

* Let's pursue traits approach + predicate instead of role variables. This will solve multiple problems for us, including trusted clusters and other use cases. Let's make sure traits rules are evaluated ahead of OIDC mapping. 
* Let's factor out SSO mapping from connector into SSO to roles mapping rules. Once we complete this, we will end up with a complete system for solving a problem of incomplete data coming from identity providers, something we are struggling with now quite often.

zmb3 commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,439 @@
+
+Variables can be used anywhere traits can be used in a role spec, including
+allow or deny rules such as `logins` or `node_labels`, impersonation conditions,
+etc.
```
What about in label keys?

nklaassen replied:
It looks like traits can be used in label keys today, but if the trait expands to a list, only the first value of the list is used (which seems very error-prone). But the var could be used the same way a trait can here.
https://github.com/gravitational/teleport/blob/fc05eaf305f32b943a0e18b3a84c9f61a544b1ae/lib/services/role.go#L446

nklaassen committed:
add alternative "predicate language for filtering only"

zmb3 commented on rfd/0078-role-variables.md:
```diff
@@ -0,0 +1,439 @@
+  name: users
+spec:
+  vars:
+    - name: allow_envs
```
So a variable can expand to a single value or a list of values, right?

But there's no difference in syntax, so it doesn't look like we can validate when used incorrectly.

What happens in this example?

- `vars.a` expands to the list containing `["1", "2"]`
- `vars.b` is:
```
out:
- "3"
- "{{vars.a}}"
```

- Is `b` "flattened" to then produce the result `["3", "1", "2"]?
- Would this also be a valid definition for b? `out: {{vars.a}}`?
- If yes, would the same hole true if `a` evaluated to a single value instead of a list?

nklaassen replied:
Yes, it will be flattened. A single value is really just a list of length one, it's not really a special case.

This is similar semantically to how traits are used today, lists are always implicitly expanded and flattened. If you have traits `groups = [a,b,c]` then
```yaml
node_labels:
   owner: ["d", "external.groups"]
```
expands to
```yaml
node_labels:
   owner: ["d", "a", "b", "c"]
```

>Would this also be a valid definition for b? `out: {{vars.a}}`
Almost, it would have to be `out: ["{{vars.a}}"]`

nklaassen committed:
update to trait transforms

nklaassen committed:
rename file

nklaassen committed:
explain new tctl test command

nklaassen committed:
mention future work on role mapping rules

nklaassen committed:
add extra example

nklaassen commented:
@klizhentas I have updated this RFD to focus on trait transforms now instead of the old role variables design. I will be OOO for the next ~week, if you get a chance to review while I'm gone I will respond when I get back.

klizhentas commented on rfd/0078-trait-transforms.md:
```diff
@@ -0,0 +1,754 @@
+which may be better names (such as `select`, `default`) must unfortunately be
+avoided.
+
+#### `list(...items)`
```
nit: flatten describes the function better in my opinion

nklaassen replied:
True, changed to `flatten`. Also added basic `list` helper which I think looks better when there is nothing to flatten, and could catch mistakes if accidentally passing a list.

klizhentas reviewed (approved):
LGTM

nklaassen committed:
s/teleport/Teleport/

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

nklaassen committed:
rename list to flatten, add actual list helper

klizhentas commented on rfd/0078-trait-transforms.md:
```diff
@@ -0,0 +1,754 @@
+  # priority can be used to order the evaluation of multiple trait transforms in
+  # a cluster. Lower priorities will be evaluated first. Trait transforms with
+  # the same priority will be ordered by a lexicographical sort by their names. 
+  priority: 0
```
I would not worry about that so early on. Otherwise we'd have to use compare and swap or other locking mechanism to prevent duplicates on create.

nklaassen replied:
yeah, good call

nklaassen replied:
Oh I hadn't seen your comment sasha, GitHub was out of date. I had updated the RFD to block dupes, but reverted now.

Compare-and-swap wouldn't be too difficult for a single resource, but across multiple resources it could be a bit of a pain.

nklaassen committed:
block duplicate priorities

nklaassen committed:
Revert "block duplicate priorities"

This reverts commit 72174b8fa1dd13184cc685db7d5fbb683e1a1da9.

klizhentas commented on rfd/0078-trait-transforms.md:
```diff
@@ -0,0 +1,762 @@
+
+Returns the local part of an email field.
+
+#### `regexp.replace(variable, expression, replacement)`
```
@nklaassen Do we need regexp.replace for now? Can we introduce a simple string replace and then evolve it? @nklaassen I will explain the rationale on the follow up call

nklaassen replied:
We can certainly go ahead without it for now, I included it because we support it for role templates currently and it seems to be a useful feature.

@klizhentas would you like me to schedule another call?

klizhentas commented on rfd/0078-trait-transforms.md:
```diff
@@ -0,0 +1,762 @@
+which may be better names (such as `select`, `default`) must unfortunately be
+avoided.
+
+#### `list(...items)`
```
@nklaassen instead of lists, I think we should operate on sets - sets imply that there are no duplicates, and adding existing value to the set does not change it. In sets order also does not matter, just like in our RBAC and trait analysis. 

What do you think?

nklaassen replied:
yes, I think working with sets makes a lot more sense

klizhentas reviewed (commented):
@nklaassen 

Sorry for the late feedback.  I was working on modeling your trait conversions in another tool and I suggest we switch to sets instead of lists and introduce a more generic form of resource, `login_rule`: here are some thoughts:

**We can remove lots of complexity**

Fflatten, specifically, could be reduced if we assume that traits are the `key: set(value)`.

What do you think about the following operations:

```python
# adds one or several values only if they does not exist yet, no need to flatten, does not add empty values
# (will be clear below)
set(external.groups, "splunk", "another-value", "") 

# contains returns true if the groups contains value splunk
contains(external.groups, "splunk") 

# instead of a separate filter function. I suppose it will be used quite frequently to only keep traits used in RBAC
# and remove the rest
keep_keys(external, "groups") # keeps only trait values "groups" if present

match(contains(external.groups, "splunk"),
        option(true, add(external.groups, "dbs")), # no need to worry about flatten, it's implied in sets
        option(false, external.groups))

# to add new traits to the cert:
set(external.logins, 
                regexp.replace(external.username, "-", "_"),
                match(external.email,
                        option("nic@goteleport.com", "root")), # this will be skipped if match failed, as it results in empty value,                
```

**We can avoid adding other resources in the future**

If the rule resource is more generic, it will spare us from adding other types of resources for role mapping,
and look like this:

```yaml
kind: login_rule
version: v1
metadata:
  name: example
spec:
  # priority can be used to order the evaluation of multiple trait transforms in
  # a cluster. Lower priorities will be evaluated first. Trait transforms with
  # the same priority will be ordered by a lexicographical sort by their names. 
  priority: 0
  expr: | 
      keep_keys(external, "groups") # keeps only trait values "groups" if present
      match(contains(external.groups, "splunk"),
        option(true, add(external.groups, "dbs")), # no need to worry about flatten, it's implied in sets
        option(false, external.groups))
      # this rule resource on sets can be used to map users to roles as well, we can avoid having too many different resources
      set(user.roles, "prod")
      keep_keys(user.roles, "non-prod")
```
 
@xacrimon this should answer your question on mappings of users to policies - policy resource will look identical to `login_rule` resource.

nklaassen commented:
@klizhentas I like the idea of working with sets over lists.

My current design with `filter` and the `override` map was mostly guided by an attempt to avoid multi-statement predicate expressions with mutable state. But I see the value in having the single expression which can do everything and be quite extensible. I will do some experimenting with it and give it some more thought.

klizhentas commented:
@nklaassen I have successfully modeled the maps in another tool. Here is what I got:

* Static map

Each login rule only contains one expression that evaluates to a map. This way it is much easier to model, test and evaluate it.
The expression can be as simple as specifying static traits:

```go
{
  'fruits': []{'apple', 'strawberry', 'banana'},
  'veggies': []{'potato'},
}
```

* To create a map with only certain arguments selected, users can refer to existing variables and other maps:

{
   'groups': external['groups'],
   'logins': []{external.email},
}

* To modify the variables and add values, users can use the `if` expression:

```go
{
            'our-fruits': If(
                external['fruits'].contains('banana'),
                external['fruits'].add('blueberry'),
                external['fruits']
            )
        }
```

You can also spot set functions `contains` and `add` used above and `if` statement. 

* And finally, users can override and remove keys from the external traits or traits specified in previous rule to chain rules:

```
external.remove_keys('logins', 'groups').overwrite({
  'logins': []{external.email}
}
```

```go
prev.traits.overwrite({
  'logins': []{external.email}
}
```

This adds the following syntax to the predicate:

* Dictionary with set-like behavior. 

```go
dict.add_value(key, value) # add single value to set it does not exists
dict['key'].contains(value) # checks if value exists and returns true if it does
dict['key'].add(value) # add value to set corresponding to key if it not exists
dict.overwrite(another_dict)
dict.remove_keys(key1, key2)
if(cond, true-case, false-case) # functional-style if else
```

russjones commented:
@nklaassen How hard would it be to add support for `lower` and `upper`? We've had a lot of requests for that functionality, see https://github.com/gravitational/teleport/issues/5267.

nklaassen commented:
> @nklaassen How hard would it be to add support for `lower` and `upper`? We've had a lot of requests for that functionality, see #5267.

It's not hard at all with the predicate language. @klizhentas would this be difficult to model in Z3?

nklaassen committed:
switch to login rules

nklaassen committed:
rename file

nklaassen committed:
mention necessary changes to predicate language

klizhentas commented:
@nklaassen I don't think this will be hard to model lower and upper.

klizhentas commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,405 @@
+
+#### `set.remove(value)`
+
+`set.add(value)` returns a copy of the set with `value` removed.
```
typo

klizhentas commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,405 @@
+```
+dict(
+  pair("fruits", set("apple")),
+).add("fruits", "banana").add("vegetables", "asparagus", "brocolli")
````
typo

klizhentas commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,405 @@
+
+#### `match(value, ...options)`
+
+`match(value, ...options)` returns the value of the first `option` which is a
```
I have implemented it slightly differently, it`s:

`select` with no argument, it simply alternates between options:

```
select {
   case 
}
```

nklaassen replied:
@klizhentas I think we can't use `select` because it's a keyword and is a statement rather than an expression in Go, the parser really doesn't like it. For now I've tried to recreate this style but named it `choose`, let me know if you think there's a better way to name it

klizhentas reviewed (changes requested):
almost there, but let's align match to be closer to select expression and fix some typos. If you want to add lower and upper initially, let's add that too.

nklaassen committed:
fix typos

nklaassen committed:
replace match with choose

nklaassen committed:
add upper and lower helpers

nklaassen committed:
remove unnecessary default_option

nklaassen committed:
drop mention of caching parsed expressions

nklaassen committed:
detail new types and RPCs

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+    // exist.
+    rpc CreateLoginRule(types.LoginRuleV1) returns (google.protobuf.Empty);
+
+    // UpsertLoginRule upserts a login rule in a backend.
+    rpc UpsertLoginRule(types.LoginRuleV1) returns (google.protobuf.Empty);
```
Usually there is `tctl create resource.yaml` to create a resource only if it does not already exist and `tctl create -f resource.yaml` to create if it does not exist or overwrite if it does, and these seem to typically have `Create` and `Upsert` endpoints. Any reason to prefer `Update`?

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+The predicate language currently does not support "methods" on objects such as
+`dict.remove_keys` or `set.contains` as described above.
+We will need to add support for these in our
+[gravitational/predicate](https://github.com/gravitational/predicate)
```
I'm not really sure - @kimlisa do we plan/have we tried to upstream your changes in the fork? Would need to got those and mine upstreamed before we could potentially get off the fork

kimlisa replied:
> Are we going to be maintaining this fork indefinitely?

yes, i believe we are, @klizhentas wanted us to fork this (he was the original creator of `vulcand/predicate` btw), so we can have our own engineers review our additions and add more functionality pertaining to our use.

klizhentas commented:
@nklaassen I'm looking at this RFD now with a fresh eye, and functional format:

```
map(
    key(...
```

Looks a bit too alien for what users are used to now.  Can we support simplified yaml format

```yaml
traits:
    key: [val]
```

that will be a shortcut to:

```yaml
traits: |
     map(key, set("val"))
```

cc @hugoShaka

vitorenesduarte commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+`choose(option(set("a").contains("b"), "foo"), option(set("a").contains("a"), "bar"))`
+returns `"bar"`.
+
+Use an option with the condition hardcoded to `true` to set a default value.
```
Would it make sense to call it `case` instead of `choose`?

nklaassen replied:
`case` is a Go keyword which is not easily handled by the parser

vitorenesduarte commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+
+#### `choose(...options)`
+
+`choose(...options)` returns the value of the first `option` for which the
```
IDK if it's useful to mention somewhere that 

`choose(option(cond, value_if_true), option(true, value_if_false))`

is equivalent to

`ifelse(cond, value_if_true, value_if_false)`

and that `choose` could be written as a chain of `ifelse`s.

vitorenesduarte commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,504 @@
+  "trees": ("aspen"),
+}
+```
+
````
We could also have a `dict.contains` method to mimic the `set` type.

nklaassen replied:
if there is a use case we could add it

vitorenesduarte reviewed (commented):
I apologize for the unrequested review, but I came across the RFD and had some questions/suggestions while reading it.

nklaassen committed:
update gRPCs

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+    rpc UpsertLoginRule(types.LoginRuleV1) returns (google.protobuf.Empty);
+
+    // DeleteLoginRule deletes an existing login rule in the backend.
+    rpc DeleteToken(types.ResourceRequest) returns (google.protobuf.Empty);
```
this is common in the legacy grpc package, updated to a new grpc service and unshared types

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+  name: example
+spec:
+  # priority can be used to order the evaluation of multiple login rules within
+  # a cluster. Lower priorities will be evaluated first. Login rules with the
```
> It's common for integers to default to zero, making those rules jump to the top of the priority list by default

Where is the top? It's not clear. A priority 0 rule will be evaluated first, but its output will be completely overwritten by a rule at priority 1, the latter p1 rule gets the final say.

codingllama replied:
If, as you said, priority is a dubious concept, should we call it "evaluation order" instead?

Whatever you want to name it, if we go with an int field then it will default to zero when not explicitly defined, which is something to consider.

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+Teleport roles which the user should have.
+
+During login, the auth server will load all `login_rule` resources in the
+cluster, sort them by `priority` and `name`, and apply all of them in order.
```
Priority *is* the way users interact with the rule-sorting system, and sorting by name *is* the explicit untie mechanism - I don't understand the concern. Do you think name is a bad untie mechanism and there is a better one?

codingllama replied:
My concern is that using names to untie may hide unintended priority conflicts, but I suppose it's a simple way to get to a working system.

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+`set.remove(value)` returns a copy of the set with `value` removed.
+`set("a", "b", "c").remove("c")` returns `("a", "b")`.
+
+#### `dict(...pairs)`
```
they're always sets

nklaassen committed:
minor syntax updates and clarifications

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+Duplicates are not possible and will be filtered out on creation of the set if
+they are passed in.
+All items must be strings.
+As a special case, the empty string will not be included in the set.
```
removed this special case, you can always call `.remove("")`

nklaassen commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,504 @@
+}
+```
+
+Arguments must be pairs, the first element will be the key,
````
you will get an error when the expression is evaluated, added a mention

nklaassen commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,504 @@
+```
+dict(
+  pair("fruits", set("apple")),
+).add_values("fruits", "banana").add("vegetables", "asparagus", "brocolli")
````
I think `append_to_key` is confusing as well, `append` is not the operation on the set, it's `add`, and you are adding values, so I think this name is okay?

codingllama replied:
Fair enough  👍

nklaassen commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,504 @@
+}
+```
+
+#### `dict.remove_keys(...keys)`
````
switch to just `remove` but imo it might as well support multiple keys. A known customer usecase will be removing many traits, `remove("a", "b", "c")` seems nicer than `.remove("a").remove("b").remove("c")`

codingllama replied:
Sounds good.

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+`input` can be a string or a set of strings, in which case the replacement will
+be applied to all strings in the set.
+
+#### `upper(input)`
```
We do really want the transformation, this is meant to set the users traits, usually the comparison happens later on with allow/deny labels in roles

I think I prefer `upper` to `toupper` because it sounds more declarative than imperative

nklaassen commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+before Teleport maps these to internal Teleport roles via `attributes_to_roles` or
+`claims_to_roles` so that transformed traits can be used for this mapping.
+
+An eventual extension to login rules could also return the desired set of
```
I added a bit more of a blurb on this, but I don't want to spend too much time fully considering it now, we're intentionally leaving it out of this rfd and scope of work

nklaassen commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,504 @@
+}
+```
+
+Arguments must be pairs, the first element will be the key,
````
It seems like something to avoid, but in practice the last one will override the previous

nklaassen commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,504 @@
+```
+dict(
+  pair("fruits", set("apple")),
+).add_values("fruits", "banana").add("vegetables", "asparagus", "brocolli")
````
typo, fixed, thanks

nklaassen committed:
add traits_map

nklaassen committed:
update proto spec

nklaassen committed:
add section on predicate expression context

nklaassen commented on rfd/0078-login-rules.md:
````diff
@@ -0,0 +1,504 @@
+`dict()` returns a dictionary with keys of type `string` and values of type `set`.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+)
````
There's prior art in racket where a list of pairs is a dict, and arguably the classic `m := { "a": "b", "c": "d" }` is creating a dict from a list of pairs as well, we just have curly braces and colons for syntactic sugar.

A pair of lists can be useful in some cases but I don't think it's the right way to express this.

We could do away with the explicit `pair()` calls and just have `dict(key, val, key, val)` like you said, it's not much different, I'm not really sure which I'd consider simpler (it's less explicit and you could get the indexing wrong).

I'll see if anyone else has a preference.

klizhentas replied:
I'd stay with `pair` for now, as dict is a collection of tuples in python as well as in racket

nklaassen committed:
protobuf typo fixes

klizhentas commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,704 @@
+
+### Why both `traits_map` and `traits_expression`
+
+`traits_expression` is meant to be an extremely powerful way for users with
```
style nit from "On writing well". If you remove "meant to be an extremely powerful" and "be able to" the sentence will still work:

`traits_expression` is a way for users with complex requirements to express their traits. It has the ...

nklaassen replied:
Yeah that is much more succinct, I like it. Thanks for the quick review!

klizhentas reviewed (approved):
Looks great, let's :shipit:

nklaassen committed:
writing suggestions

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,703 @@
+message LoginRule {
+  // kind is a resource kind
+  string kind = 1;
+  // sub_kind is an optional resource sub kind, used in some resources
```
Is this relevant for LoginRules? Could we clearly state it, in any case?

nklaassen replied:
not relevant. I will state it or remove this if possible, just copied a generic comment here that is on almost all resources

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,703 @@
+// LoginRule is a resource to configure rules and logic which should run during
+// Teleport user login.
+message LoginRule {
+  // kind is a resource kind
```
What is the kind for LoginRules? Does it support a multitude of kinds or only one?

I understand the fields here are added to make LoginRules conform to the "resources" interface, but I would consider alternatives to be conformant without adding superficial fields to the gRPC/proto APIs.

nklaassen replied:
the kind will be `login_rule` as it's shown in all the YAML examples. That's the only kind relevant for this type, it's used for YAML parsing, I'll see about keeping that out of the protobufs

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,703 @@
+  string kind = 1;
+  // sub_kind is an optional resource sub kind, used in some resources
+  string sub_kind = 2;
+  // version is a resource version
```
Similarly to above, what are the current/support versions? Let's document.

nklaassen replied:
it's going to start with v1, maybe one day there will be a v2... I can document if I end up keeping this here

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,703 @@
+  // priority is the priority of the login rule relative to other login rules
+  // in the same cluster. Login rules with a lower numbered priority will be
+  // evaluated first.
+  string priority = 1;
```
Did you pick string to avoid the zero value of zero? I'm concerned a string field allows for many more invalid inputs than an integer.

nklaassen replied:
no that's a mistake/typo, it should be some integer type, I'll fix

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,703 @@
+  // CreateLoginRule creates a login rule if one with the same name does not
+  // already exist, else it returns an error.
+  // (RFD note) Used for: tctl create rule.yaml
+  rpc CreateLoginRule(CreateLoginRuleRequest) returns (CreateLoginRuleResponse);
```
Suggestion:

```suggestion
  rpc CreateLoginRule(CreateLoginRuleRequest) returns (LoginRule);
```

https://cloud.google.com/apis/design/standard_methods#create

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,703 @@
+  // UpsertLoginRule creates a login rule if one with the same name does not
+  // already exist, else it replaces the existing login rule.
+  // (RFD note) Used for: tctl create -f rule.yaml
+  rpc UpsertLoginRule(UpsertLoginRuleRequest) returns (UpsertLoginRuleResponse);
```
Suggestion:

```suggestion
  rpc UpsertLoginRule(UpsertLoginRuleRequest) returns (LoginRule);
```

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,703 @@
+
+## Required Approvers
+
+- @klizhentas
```
Should we add Security approvers? Engineering? I assume Sasha is for product, but should we CC Xin as well?

nklaassen replied:
I listed Sasha here because he proposed these ideas and worked pretty closely on this. Is this a suggestion to add more people to the list? Is there a policy of who to include here? I don't know.

codingllama replied:
I've been following https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md#required-approvers. There isn't a fixed policy that I know of, I tend double-check with my manager/stakeholders and try to cover the 3 areas.

codingllama commented on rfd/0078-login-rules.md:
```diff
@@ -0,0 +1,504 @@
+  // version is a resource version
+  string version = 3;
+  // metadata is resource metadata
+  Metadata metadata = 4 [(gogoproto.nullable) = false];
```
Yeah, I stumbled on this myself since the comment. It's a sad tie-in with Gogo in all of our events/resources.

nklaassen committed:
use int32 for priority

nklaassen committed:
update rpc return types

nklaassen committed:
proto updates

nklaassen committed:
Merge branch 'master' into rfd/0078-role-variables
